File: OS X/TPLoginWindowController.m
#import "TPLoginWindowController.h"

@implementation TPLoginWindowController

- (id) init
{
	self = [super initWithWindowNibName:@"TPLoginWindowController"];
	if (self)
	{
	}
	return self;	
}

- (void)awakeFromNib
{
	if (loginNameTemp) [loginNameField setStringValue:loginNameTemp];
	if (passwordTemp) [passwordField setStringValue:passwordTemp];	
	
	if (loginNameTemp)
		[self.window makeFirstResponder:passwordField];
	else
		[self.window makeFirstResponder:loginNameField];
}

- (void) dealloc
{
	[loginNameTemp release];
	[passwordTemp release];
	[parentWindow release];
	[super dealloc];
}

- (NSString *) loginName
{
	return [loginNameField stringValue];
}

- (NSString *) password
{
	return [passwordField stringValue];
}

- (void) setLoginName:(NSString *)aLoginName
{
	loginNameTemp = [aLoginName retain];
	[loginNameField setStringValue:aLoginName];
}

- (void) setPassword:(NSString *)aPassword
{
	passwordTemp = [aPassword retain];
	[passwordField setStringValue:aPassword];
}

- (IBAction) onLoginClick:(id)sender
{
	done = YES;
	[NSApp stopModal];
}

- (IBAction) onCancelClick:(id)sender
{
	done = NO;
	[NSApp stopModal];
}

- (BOOL) promptForLogin:(NSWindow *)aParentwindow
{
	done = NO;
	parentWindow = [aParentwindow retain];
	
	if ([aParentwindow isVisible])
	{
		[NSApp beginSheet:[self window] modalForWindow:aParentwindow modalDelegate:self didEndSelector:@selector(didEndSheet:returnCode:contextInfo:) contextInfo:nil];
		@try
		{
			[NSApp runModalForWindow: [self window]];
		}
		@finally
		{
			[NSApp endSheet:[self window]];	
			[[self window] orderOut:parentWindow];
		}
	}
	else
	{
		NSModalSession s = [NSApp beginModalSessionForWindow:[self window]];
		[[self window] display];
		@try
		{
			[NSApp runModalForWindow: [self window]];
		}
		@finally 
		{
			[NSApp endModalSession:s];
			[[self window] close];
		}
	}
	
	return done;
}

- (void)didEndSheet:(NSWindow *)sheet returnCode:(int)returnCode contextInfo:(void *)contextInfo
{
    [[self window] orderOut:parentWindow];
}

@end

File: OS X (C#)/TPLoginWindowController.cs
using AppKit;

namespace RemObjects.TwinPeaks.OSX
{
	public class TPLoginWindowController: NSWindowController
	{
		public override instancetype init()
		{
			this = base.initWithWindowNibName("TPLoginWindowController");
			if (this != null)
			{
			}
			return this;
		}

		public override void awakeFromNib()
		{
			if (loginNameTemp != null)
				loginNameField.setStringValue(loginNameTemp);
			if (passwordTemp != null)
				passwordField.setStringValue(passwordTemp);
			if (loginNameTemp != null)
				this.window.makeFirstResponder(passwordField);
			else
				this.window.makeFirstResponder(loginNameField);
		}

		private void didEndSheet(NSWindow sheet) returnCode(int returnCode) contextInfo(void* contextInfo)
		{
			this.window().orderOut(parentWindow);
		}

		protected NSTextField loginNameField;
		protected NSTextField passwordField;
		protected bool done;
		protected NSWindow parentWindow;
		protected NSString loginNameTemp;
		protected NSString passwordTemp;

		protected NSString loginName
		{
			get
			{
				return loginNameField.stringValue();
			}

			set
			{
				loginNameTemp = value;
				loginNameField.setStringValue(value);
			}
		}

		protected NSString password
		{
			get
			{
				return passwordField.stringValue();
			}

			set
			{
				passwordTemp = value;
				passwordField.setStringValue(value);
			}
		}

		public void onLoginClick(id sender)
		{
			done = true;
			NSApp.stopModal();
		}

		public void onCancelClick(id sender)
		{
			done = false;
			NSApp.stopModal();
		}

		public bool promptForLogin(NSWindow aParentwindow)
		{
			done = false;
			parentWindow = aParentwindow;
			if (aParentwindow.isVisible())
			{
				NSApp.beginSheet(this.window()) modalForWindow(aParentwindow) modalDelegate(this) didEndSelector(__selector(didEndSheet:returnCode:contextInfo:)) contextInfo(null);
				try
				{
					NSApp.runModalForWindow(this.window());
				}
				finally
				{
					NSApp.endSheet(this.window());
					this.window().orderOut(parentWindow);
				}
			}
			else
			{
				NSModalSession s = NSApp.beginModalSessionForWindow(this.window());
				this.window().display();
				try
				{
					NSApp.runModalForWindow(this.window());
				}
				finally
				{
					NSApp.endModalSession(s);
					this.window().close();
				}
			}
			return done;
		}
	}
}
